File: roll.pl
#!/usr/bin/env perl

use strict;

#   .
#  / \
# / 4 \
# `---´
sub printd4 {
    my ($d) = @_;
    print "\n  .\n / \\\n/ $d \\\n`---´\n";
}

# .---.
# | 6 |
# '---'
sub printd6 {
    my ($d) = @_;
    print "\n.---.\n| $d |\n'---'\n";
}

#  /'\
# /___\
# \ 8 /
#  \./
sub printd8 {
    my ($d) = @_;
    print "\n /'\\\n/___\\\n\\ $d /\n \\./\n";
}

#  ./\.
# //10\\
# ``--´´
sub printd10 {
    my ($d) = @_;
    printf "\n ./\\. \n//%02d\\\\\n``--´´\n", $d;
}

#  .__.
# /\__/\
# \/12\/
#  `--´
sub printd12 {
    my ($d) = @_;
    printf "\n .__. \n/\\__/\\\n\\/%02d\\/\n `--´ \n", $d;
}

#   __
#  /__\
# /\20/\
# \_\/_/
sub printd20 {
    my ($d) = @_;
    printf "\n  __  \n /__\\ \n/\\%02d/\\\n\\_\\/_/\n", $d;
}

#   .
# .´ `.
# \100/
#  `-´
sub printd100 {
    my ($d) = @_;
    printf "\n  .\n.´ `.\n\\%03d/\n `-´\n", $d;
}

sub parsearg {
    my ($arg) = @_;
    return split('d', $arg);
    #my ($cnt, $die) = split('d', $arg);
    #return ($cnt, $die);
}

sub processroll {
    my ($cnt, $die) = @_;

    my $printdice;
    if ($die == 4) {
        $printdice = \&printd4;
    }
    elsif ($die == 6) {
        $printdice = \&printd6;
    }
    elsif ($die == 8) {
        $printdice = \&printd8;
    }
    elsif ($die == 10) {
        $printdice = \&printd10;
    }
    elsif ($die == 12) {
        $printdice = \&printd12;
    }
    elsif ($die == 20) {
        $printdice = \&printd20;
    }
    elsif ($die == 100) {
        $printdice = \&printd100;
    }
    else {
        # error, not a valid dice
        return 0;
    }

    my $result = 0;
    my $sum = 0;
    for (1..$cnt) {
        $result = int(rand($die))+1;
        &$printdice($result);
        $sum += $result;
    }

    return $sum;
}

# main
my $sum = 0;
my $mods = 0;
foreach my $arg (@ARGV) {
    if ($arg =~ /d/) {
        my ($cnt, $die) = parsearg($arg);
        $sum += processroll($cnt, $die);
    }
    elsif ($arg =~ /^(-|\+)/) {
        $mods += int($arg);
    }
}

print "\nTotal modifiers: $mods\n";
$sum += $mods;
        
print "\n$sum\n";

File: roll.lua
#!/usr/bin/env lua

printf = function(s,...) return io.write(s:format(...)) end

--   .
--  / \
-- / 4 \
-- `---´
function printd4(d)
  printf("\n  .\n / \\\n/ %d \\\n`---´\n", d)
end

-- .---.
-- | 6 |
-- '---'
function printd6(d)
  printf("\n.---.\n| %d |\n'---'\n", d)
end

--  /'\
-- /___\
-- \ 8 /
--  \./
function printd8(d)
  printf("\n /'\\\n/___\\\n\\ %d /\n \\./\n", d)
end

--  ./\.
-- //10\\
-- ``--´´
function printd10(d)
  printf("\n ./\\. \n//%02d\\\\\n``--´´\n", d)
end

--  .__.
-- /\__/\
-- \/12\/
--  `--´
function printd12(d)
  printf("\n .__. \n/\\__/\\\n\\/%02d\\/\n `--´ \n", d)
end

--   __
--  /__\
-- /\20/\
-- \_\/_/
function printd20(d)
  printf("\n  __  \n /__\\ \n/\\%02d/\\\n\\_\\/_/\n", d)
end

--   .
-- .´ `.
-- \100/
--  `-´
function printd100(d)
  printf("\n  .\n.´ `.\n\\%03d/\n `-´\n", d)
end

function parseroll(arg)
  local i, j, cnt, die
  i, j = string.find(arg, 'd')
  cnt = (i > 1 and string.sub(arg, 1, i-1) or 1)
  die = string.sub(arg, j+1, #arg)
  return tonumber(cnt), tonumber(die)
end

function processroll(cnt, die)
  local printdice
  if die == 4 then
    printdice = printd4
  elseif die == 6 then
    printdice = printd6
  elseif die == 8 then
    printdice = printd8
  elseif die == 10 then
    printdice = printd10
  elseif die == 12 then
    printdice = printd12
  elseif die == 20 then
    printdice = printd20
  elseif die == 100 then
    printdice = printd100
  else
    -- error, not a valid dice
    return 0
  end

  local sum, roll = 0, 0
  for i=1,cnt do
    roll = math.random(die)
    printdice(roll)
    sum = sum + roll
  end
  return sum
end

-- main
sum = 0
for i,v in ipairs(arg) do
  cnt, die = parseroll(v)
  sum = sum + processroll(cnt, die)
end
printf("\n%d\n", sum)
